File: app/orders/orderService.py
import uuid
import hashlib
from datetime import datetime, timedelta
from ..database import db
from .models import Order, OrderItem, Payment, DigitalDownload, Cart, CartItem
from ..books.models import Book

# Logic for Payment validation & DRM
def validate_payment(order_id):
    # Retrieve order and payment details
    payment = Payment.query.filter_by(OrderID=order_id).first()
    if not payment:
        return {"error": "Payment record not found", "status": 404}
    return {"message": "Payment verified", "payment_status": payment.PaymentStatus, "status": 200}

def manage_drm(order_item_id):
    # Generates or fetches a DRM link for an eBook
    download = DigitalDownload.query.filter_by(OrderItemID=order_item_id).first()
    if not download:
        return {"error": "Download link not found", "status": 404}
    
    if download.ExpiryDate < datetime.utcnow():
        return {"error": "Download link expired", "status": 403}
        
    return {"download_url": download.DownloadLink, "status": 200}

def _mock_payment_gateway(payment_info):
    """
    Simulates a payment gateway response.
    Returns a success token if credit card is not '0000', otherwise fails.
    """
    card_number = payment_info.get("card_number", "")
    if card_number == "0000000000000000":
        return {"status": "Fail", "message": "Insufficient funds"}
    return {"status": "Success", "transaction_id": f"txn_{uuid.uuid4().hex[:10]}"}

def _get_or_create_cart(user_id):
    """Internal helper to get or create a user's cart"""
    cart = Cart.query.filter_by(UserID=user_id).first()
    if not cart:
        cart = Cart(UserID=user_id)
        db.session.add(cart)
        db.session.commit()
    return cart

def get_user_cart(user_id):
    """Fetches the user's cart and items with calculated totals"""
    cart = _get_or_create_cart(user_id)
    cart_items = CartItem.query.filter_by(CartID=cart.CartID).all()
    
    items_data = []
    total_amount = 0
    
    for item in cart_items:
        book = Book.query.get(item.BookID)
        if book:
            subtotal = book.Price * item.Quantity
            total_amount += subtotal
            items_data.append({
                "book_id": book.BookID,
                "title": book.Title,
                "price": str(book.Price),
                "quantity": item.Quantity,
                "subtotal": str(subtotal),
                "format": book.Format,
                "cover_image": book.CoverImageURL
            })
            
    return {
        "cart_id": cart.CartID,
        "items": items_data,
        "total_amount": str(total_amount),
        "item_count": sum(item.Quantity for item in cart_items)
    }

def add_to_cart(user_id, book_id, quantity=1):
    """Adds a book to the cart or increments quantity"""
    book = Book.query.get(book_id)
    if not book:
        return {"error": "Book not found", "status": 404}
        
    cart = _get_or_create_cart(user_id)
    
    # Check if item already exists
    cart_item = CartItem.query.filter_by(CartID=cart.CartID, BookID=book_id).first()
    
    if cart_item:
        cart_item.Quantity += quantity
    else:
        cart_item = CartItem(CartID=cart.CartID, BookID=book_id, Quantity=quantity)
        db.session.add(cart_item)
        
    db.session.commit()
    return {"message": f"Added {book.Title} to cart", "status": 200}

def update_cart_item(user_id, book_id, quantity):
    """Updates the quantity of a cart item. If 0, removes it."""
    cart = _get_or_create_cart(user_id)
    cart_item = CartItem.query.filter_by(CartID=cart.CartID, BookID=book_id).first()
    
    if not cart_item:
        return {"error": "Item not in cart", "status": 404}
        
    if quantity <= 0:
        db.session.delete(cart_item)
        msg = "Item removed from cart"
    else:
        cart_item.Quantity = quantity
        msg = "Cart updated"
        
    db.session.commit()
    return {"message": msg, "status": 200}

def clear_user_cart(user_id):
    """Empties the user's cart"""
    cart = Cart.query.filter_by(UserID=user_id).first()
    if cart:
        CartItem.query.filter_by(CartID=cart.CartID).delete()
        db.session.commit()
    return {"message": "Cart cleared", "status": 200}

def process_checkout(user_id, shipping_address, payment_info):
    """
    MODIFIED: Now pulls items directly from the database Cart.
    """
    cart = Cart.query.filter_by(UserID=user_id).first()
    if not cart or not cart.items:
        return {"error": "Your cart is empty", "status": 400}

    # 1. Validation & Inventory Check
    total_amount = 0
    book_objects = []
    
    for item in cart.items:
        book = Book.query.get(item.BookID)
        if not book:
            continue
            
        if book.Format == 'Physical' and book.StockQuantity < item.Quantity:
            return {"error": f"Insufficient stock for {book.Title}. Only {book.StockQuantity} left.", "status": 400}
            
        total_amount += (book.Price * item.Quantity)
        book_objects.append({"book": book, "quantity": item.Quantity})

    if not book_objects:
         return {"error": "Cart contains invalid books", "status": 400}

    # 2. Financial Transaction (Dummy)
    payment_response = _mock_payment_gateway(payment_info)
    if payment_response["status"] == "Fail":
        return {"error": payment_response["message"], "status": 402}

    # 3. Execution Flow
    try:
        # Create Order
        new_order = Order(
            UserID=user_id,
            TotalAmount=total_amount,
            ShippingAddress=shipping_address,
            OrderStatus='Pending'
        )
        db.session.add(new_order)
        db.session.flush()

        # Payment Record
        new_payment = Payment(
            OrderID=new_order.OrderID,
            PaymentMethod=payment_info.get("method", "Credit Card"),
            TransactionID=payment_response["transaction_id"],
            PaymentStatus='Success',
            PaymentDate=datetime.utcnow(),
            Amount=total_amount
        )
        db.session.add(new_payment)

        # Process Items
        has_physical = False
        has_digital = False
        
        for item_data in book_objects:
            book = item_data["book"]
            quantity = item_data["quantity"]
            
            order_item = OrderItem(
                OrderID=new_order.OrderID,
                BookID=book.BookID,
                Quantity=quantity,
                UnitPrice=book.Price
            )
            db.session.add(order_item)
            db.session.flush()
            
            if book.Format == 'Physical':
                book.StockQuantity -= quantity
                db.session.add(book)
                has_physical = True
            elif book.Format in ['E-book', 'Audiobook']:
                unique_hash = hashlib.sha256(f"{user_id}_{book.BookID}_{uuid.uuid4()}".encode()).hexdigest()
                download = DigitalDownload(
                    OrderItemID=order_item.OrderItemID,
                    DownloadLink=f"/api/orders/download/{unique_hash}",
                    ExpiryDate=datetime.utcnow() + timedelta(days=365)
                )
                db.session.add(download)
                has_digital = True

        # 4. Clear the Cart after SUCCESS
        CartItem.query.filter_by(CartID=cart.CartID).delete()

        db.session.commit()
        
        # Prepare Response
        response_data = {
            "message": "Order placed successfully!",
            "order_id": new_order.OrderID,
            "transaction_id": payment_response["transaction_id"],
            "total_amount": str(total_amount)
        }
        
        if has_physical:
            response_data["estimated_delivery"] = (datetime.utcnow() + timedelta(days=5)).strftime("%Y-%m-%d")
        if has_digital:
            response_data["digital_fulfillment"] = "Check your history for downloads."
            
        return {**response_data, "status": 201}

    except Exception as e:
        db.session.rollback()
        return {"error": f"Internal server error: {str(e)}", "status": 500}

def get_user_order_history(user_id):
    """
    Fetches the order history for a specific user, including items
    and download links for digital products.
    """
    orders = Order.query.filter_by(UserID=user_id).order_by(Order.OrderDate.desc()).all()
    history = []
    
    for order in orders:
        items_data = []
        payment = Payment.query.filter_by(OrderID=order.OrderID).first()
        order_items = OrderItem.query.filter_by(OrderID=order.OrderID).all()
        
        for item in order_items:
            book = Book.query.get(item.BookID)
            item_dict = {
                "book_id": item.BookID,
                "title": book.Title if book else "Unknown",
                "format": book.Format if book else "Unknown",
                "quantity": item.Quantity,
                "unit_price": str(item.UnitPrice)
            }
            
            # Fetch download link if digital
            if book and book.Format in ['E-book', 'Audiobook']:
                download = DigitalDownload.query.filter_by(OrderItemID=item.OrderItemID).first()
                if download:
                    item_dict["download_link"] = download.DownloadLink
                    
            items_data.append(item_dict)
            
        history.append({
            "order_id": order.OrderID,
            "date": order.OrderDate.strftime("%Y-%m-%d %H:%M:%S") if order.OrderDate else None,
            "status": order.OrderStatus,
            "total_amount": str(order.TotalAmount),
            "payment_status": payment.PaymentStatus if payment else "Pending",
            "items": items_data
        })
        
    return {"items": history, "status": 200}
    
def get_all_orders():
    """Admin: Fetches all orders"""
    orders = Order.query.order_by(Order.OrderDate.desc()).all()
    all_orders = []
    
    for order in orders:
        all_orders.append({
            "order_id": order.OrderID,
            "user_id": order.UserID,
            "date": order.OrderDate.strftime("%Y-%m-%d %H:%M:%S") if order.OrderDate else None,
            "status": order.OrderStatus,
            "total_amount": str(order.TotalAmount)
        })
    return {"orders": all_orders, "status": 200}

def update_order_status(order_id, new_status):
    """Admin: Updates the fulfillment status of an order."""
    allowed_statuses = ['Pending', 'Shipped', 'Delivered', 'Cancelled']
    if new_status not in allowed_statuses:
        return {"error": f"Invalid status. Allowed: {allowed_statuses}", "status": 400}
        
    order = Order.query.get(order_id)
    if not order:
        return {"error": "Order not found", "status": 404}
        
    order.OrderStatus = new_status
    db.session.commit()
    
    return {"message": f"Order {order_id} status updated to {new_status}", "status": 200}

File: php_server/src/Orders/OrderService.php
<?php

namespace Orders;

use Core\Database;
use Exception;
use PDO;

class OrderService {

    // ============================================================
    // PRIVATE CART HELPERS
    // Cart items are stored as JSON in the `carts` table.
    // Format: [{"book_id": 1, "quantity": 2}, ...]
    // ============================================================

    /**
     * Returns raw cart items array for a user (from JSON column).
     */
    private function getRawCartItems(int $userId): array {
        $db = Database::get();
        $stmt = $db->prepare("SELECT items FROM carts WHERE user_id = ?");
        $stmt->execute([$userId]);
        $row = $stmt->fetch();
        if (!$row || $row['items'] === null) return [];
        return json_decode($row['items'], true) ?? [];
    }

    /**
     * Persists cart items array back to the JSON column (upsert).
     */
    private function saveCartItems(int $userId, array $items): void {
        $db = Database::get();
        $json = json_encode(array_values($items)); // re-index array
        $stmt = $db->prepare("
            INSERT INTO carts (user_id, items)
            VALUES (?, ?)
            ON DUPLICATE KEY UPDATE items = VALUES(items), updated_at = NOW()
        ");
        $stmt->execute([$userId, $json]);
    }

    // ============================================================
    // CART METHODS
    // ============================================================

    /**
     * Get the user's cart with full book details joined in.
     */
    public function getUserCart(int $userId): array {
        $db = Database::get();
        $rawItems = $this->getRawCartItems($userId);

        if (empty($rawItems)) {
            return ['items' => [], 'total_amount' => '0', 'item_count' => 0];
        }

        // Fetch book details for all items in one query
        $bookIds      = array_column($rawItems, 'book_id');
        $placeholders = implode(',', array_fill(0, count($bookIds), '?'));
        $stmt         = $db->prepare("SELECT id, title, price, isbn FROM books WHERE id IN ({$placeholders})");
        $stmt->execute($bookIds);

        $booksById = [];
        foreach ($stmt->fetchAll() as $book) {
            $booksById[$book['id']] = $book;
        }

        $items = [];
        $total = 0.0;

        foreach ($rawItems as $cartItem) {
            $book = $booksById[$cartItem['book_id']] ?? null;
            if (!$book) continue; // skip if book was deleted

            $total += (float) $book['price'] * $cartItem['quantity'];
            $items[] = [
                'BookID'   => $cartItem['book_id'],
                'Title'    => $book['title'],
                'Price'    => (string) $book['price'],
                'ISBN'     => $book['isbn'],
                'Quantity' => $cartItem['quantity'],
            ];
        }

        return [
            'items'        => $items,
            'total_amount' => (string) $total,
            'item_count'   => count($items),
        ];
    }

    /**
     * Add a book to cart. Validates stock before adding.
     */
    public function addToCart(int $userId, int $bookId, int $quantity): bool {
        $db = Database::get();

        // Stock validation
        $stmt = $db->prepare("SELECT stock_quantity, title FROM books WHERE id = ?");
        $stmt->execute([$bookId]);
        $book = $stmt->fetch();
        if (!$book) throw new Exception("Book not found");

        if ($book['stock_quantity'] < $quantity) {
            throw new Exception("Only {$book['stock_quantity']} units of '{$book['title']}' are available in stock.");
        }

        $items = $this->getRawCartItems($userId);
        $found = false;

        foreach ($items as &$item) {
            if ((int)$item['book_id'] === $bookId) {
                $newQty = $item['quantity'] + $quantity;
                if ($book['stock_quantity'] < $newQty) {
                    throw new Exception("Cannot add more. Total in cart ({$newQty}) exceeds available stock ({$book['stock_quantity']}).");
                }
                $item['quantity'] = $newQty;
                $found = true;
                break;
            }
        }
        unset($item);

        if (!$found) {
            $items[] = ['book_id' => $bookId, 'quantity' => $quantity];
        }

        $this->saveCartItems($userId, $items);
        return true;
    }

    /**
     * Update quantity of a cart item. quantity=0 removes the item.
     */
    public function updateCartItem(int $userId, int $bookId, int $quantity): bool {
        $db = Database::get();

        $items = $this->getRawCartItems($userId);

        if ($quantity <= 0) {
            // Remove item
            $items = array_filter($items, fn($i) => (int)$i['book_id'] !== $bookId);
        } else {
            // Stock validation
            $stmt = $db->prepare("SELECT stock_quantity FROM books WHERE id = ?");
            $stmt->execute([$bookId]);
            $stock = $stmt->fetchColumn();
            if ($stock !== false && $stock < $quantity) {
                throw new Exception("Requested quantity ($quantity) exceeds available stock ($stock).");
            }

            $found = false;
            foreach ($items as &$item) {
                if ((int)$item['book_id'] === $bookId) {
                    $item['quantity'] = $quantity;
                    $found = true;
                    break;
                }
            }
            unset($item);

            if (!$found) {
                $items[] = ['book_id' => $bookId, 'quantity' => $quantity];
            }
        }

        $this->saveCartItems($userId, $items);
        return true;
    }

    /**
     * Empty the user's cart.
     */
    public function clearCart(int $userId): bool {
        $this->saveCartItems($userId, []);
        return true;
    }

    // ============================================================
    // ORDER METHODS
    // ============================================================

    /**
     * Checkout: validates stock, creates order with embedded items
     * and payment info, decrements stock, clears cart.
     */
    public function checkout(int $userId, string $shippingAddress, array $paymentInfo): array {
        $db   = Database::get();
        $cart = $this->getUserCart($userId);

        if (empty($cart['items'])) {
            throw new Exception("Cart is empty");
        }

        $db->beginTransaction();
        try {
            $orderItems = [];

            foreach ($cart['items'] as $item) {
                // Final stock check with row lock
                $stmt = $db->prepare("SELECT stock_quantity FROM books WHERE id = ? FOR UPDATE");
                $stmt->execute([$item['BookID']]);
                $currentStock = $stmt->fetchColumn();

                if ($currentStock < $item['Quantity']) {
                    throw new Exception("Stock changed for '{$item['Title']}'. Only {$currentStock} available.");
                }

                // Decrement stock
                $stmt = $db->prepare("UPDATE books SET stock_quantity = stock_quantity - ? WHERE id = ?");
                $stmt->execute([$item['Quantity'], $item['BookID']]);

                // Build order items snapshot (title captured at time of order)
                $orderItems[] = [
                    'BookID'    => $item['BookID'],
                    'Title'     => $item['Title'],
                    'Quantity'  => $item['Quantity'],
                    'UnitPrice' => $item['Price'],
                ];
            }

            $txnId = 'PHP_TXN_' . uniqid();

            // Create order with embedded items + payment
            $stmt = $db->prepare("
                INSERT INTO orders
                    (user_id, total_amount, shipping_address, status, payment_method, payment_status, transaction_id, items)
                VALUES
                    (?, ?, ?, 'Pending', ?, 'Success', ?, ?)
            ");
            $stmt->execute([
                $userId,
                $cart['total_amount'],
                $shippingAddress,
                $paymentInfo['method'] ?? 'COD',
                $txnId,
                json_encode($orderItems)
            ]);
            $orderId = $db->lastInsertId();

            // Clear cart
            $this->clearCart($userId);

            $db->commit();
            return [
                "message"  => "Order placed successfully",
                "OrderID"  => $orderId,
                "status"   => 201
            ];
        } catch (Exception $e) {
            $db->rollBack();
            throw $e;
        }
    }

    /**
     * Get order history for a user. Items decoded from JSON.
     */
    public function getUserOrderHistory(int $userId): array {
        $db   = Database::get();
        $stmt = $db->prepare("SELECT * FROM orders WHERE user_id = ? ORDER BY order_date DESC");
        $stmt->execute([$userId]);
        $orders = $stmt->fetchAll();

        foreach ($orders as &$order) {
            $order['OrderID']         = $order['id'];
            $order['UserID']          = $order['user_id'];
            $order['TotalAmount']     = (string) $order['total_amount'];
            $order['ShippingAddress'] = $order['shipping_address'];
            $order['OrderStatus']     = $order['status'];
            $order['OrderDate']       = $order['order_date'];
            $order['PaymentStatus']   = $order['payment_status'];
            $order['Items']           = json_decode($order['items'] ?? '[]', true) ?? [];
        }

        return $orders;
    }

    /**
     * Get all orders (admin view). Joins with users for email.
     */
    public function getAllOrders(): array {
        $db   = Database::get();
        $stmt = $db->query("
            SELECT o.*, u.email AS Email
            FROM orders o
            JOIN users u ON o.user_id = u.id
            ORDER BY o.order_date DESC
        ");
        $orders = $stmt->fetchAll();

        foreach ($orders as &$order) {
            $order['OrderID']     = $order['id'];
            $order['UserID']      = $order['user_id'];
            $order['TotalAmount'] = (string) $order['total_amount'];
            $order['OrderStatus'] = $order['status'];
            $order['OrderDate']   = $order['order_date'];
        }

        return $orders;
    }

    /**
     * Update order status (e.g., Pending → Shipped).
     */
    public function updateOrderStatus(int $orderId, string $status): bool {
        $db   = Database::get();
        $stmt = $db->prepare("UPDATE orders SET status = ? WHERE id = ?");
        return $stmt->execute([$status, $orderId]);
    }
}
